[clang] Define `ATOMIC_FLAG_INIT` correctly for C++.

chrisnc opened the pull request.

chrisnc committed:
[clang] Define `ATOMIC_FLAG_INIT` correctly for C++.

chrisnc commented:
Without this change, trying to use `ATOMIC_FLAG_INIT` in a constant initialization context yields the following:

https://godbolt.org/z/sTvMs5a95

```
<source>:3:17: error: non-constant-expression cannot be narrowed from type 'int' to 'atomic_bool' (aka '_Atomic(bool)') in initializer list [-Wc++11-narrowing]
    3 | atomic_flag x = ATOMIC_FLAG_INIT;
      |                 ^~~~~~~~~~~~~~~~
/opt/compiler-explorer/clang-trunk-20240703/lib/clang/19/include/stdatomic.h:169:28: note: expanded from macro 'ATOMIC_FLAG_INIT'
  169 | #define ATOMIC_FLAG_INIT { 0 }
      |                            ^
<source>:3:17: note: insert an explicit cast to silence this issue
    3 | atomic_flag x = ATOMIC_FLAG_INIT;
      |                 ^~~~~~~~~~~~~~~~
/opt/compiler-explorer/clang-trunk-20240703/lib/clang/19/include/stdatomic.h:169:28: note: expanded from macro 'ATOMIC_FLAG_INIT'
  169 | #define ATOMIC_FLAG_INIT { 0 }
      |                            ^
1 error generated.
Compiler returned: 1
```

chrisnc commented:
ping

chrisnc commented:
ping

AaronBallman commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -750,6 +750,8 @@ Bug Fixes in This Version
 
 - Fixed `static_cast` to array of unknown bound. Fixes (#GH62863).
 
+- Fixed the definition of ATOMIC_FLAG_INIT in stdatomic.h so it can be used in C++.
```
```suggestion
- Fixed the definition of ``ATOMIC_FLAG_INIT`` in ``<stdatomic.h>`` so it can be used in C++.
```

chrisnc replied:
Done.

AaronBallman reviewed (commented):
Thank you for catching this! Can you also add some tests to `clang/test/Headers/`? We have some coverage for stdatomic.h in C, but not in C++ (I'd recommend trying to add a `RUN` line to the .c file with `-x c++` so it's tested in C++ mode and then ensure there's sufficient coverage for `ATOMIC_FLAG_INIT`.)

chrisnc commented:
Rebased and added a test case. I confirmed that the test fails using the current definition of `ATOMIC_FLAG_INIT`:

```
error: 'expected-error' diagnostics seen but not expected:
  Line 38: non-constant-expression cannot be narrowed from type 'int' to 'atomic_bool' (aka '_Atomic(bool)') in initializer list
error: 'expected-note' diagnostics seen but not expected:
  Line 38: insert an explicit cast to silence this issue
2 errors generated.
```

AaronBallman reviewed (approved):
LGTM!

AaronBallman commented:
Btw, do you need someone to merge this on your behalf?

chrisnc commented:
@AaronBallman yes, that would be great! I'm just a contributor, not a collaborator.

AaronBallman commented:
> @AaronBallman Just FYI you can also add the Milestone to the PR directly and use the /cherry-pick command right here. This is a bit simpler.

That's good to know! I was going off the documentation at: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches because it says "issue" specifically. I'll post a patch to update those docs, thanks!

AaronBallman commented:
> > @AaronBallman Just FYI you can also add the Milestone to the PR directly and use the /cherry-pick command right here. This is a bit simpler.
> 
> That's good to know! I was going off the documentation at: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches because it says "issue" specifically. I'll post a patch to update those docs, thanks!

https://github.com/llvm/llvm-project/pull/100401
